lib/core-server/src/utils/StoryIndexGenerator.ts

```diff
@@ -240,50 +327,73 @@ export class StoryIndexGenerator {
+    const changeDocsName = 'Use `<Meta of={} name="Other Name">` to distinguish them.';
 
-  async extractStories(specifier: NormalizedStoriesSpecifier, absolutePath: Path) {
-    const relativePath = path.relative(this.options.workingDir, absolutePath);
-    const entries = [] as IndexEntry[];
-    try {
-      const importPath = slash(normalizeStoryPath(relativePath));
-      const makeTitle = (userTitle?: string) => {
-        return userOrAutoTitleFromSpecifier(importPath, specifier, userTitle);
-      };
-      const csf = await this.index(absolutePath, { makeTitle });
-      csf.stories.forEach(({ id, name, parameters }) => {
-        const base = { id, title: csf.meta.title, name, importPath };
+    if (betterEntry.type === 'story') {
+      // This shouldn't be possible
+      if (worseEntry.type === 'story')
+        throw new Error(`Duplicate stories with id: ${betterEntry.id}`);
 
-        if (parameters?.docsOnly) {
-          if (this.options.docs.enabled) {
-            entries.push({ ...base, type: 'docs', storiesImports: [], legacy: true });
-          }
-        } else {
-          entries.push({ ...base, type: 'story' });
-        }
-      });
-    } catch (err) {
-      if (err.name === 'NoMetaError') {
-        logger.info(`💡 Skipping ${relativePath}: ${err}`);
+      const worseDescriptor = worseEntry.standalone
+        ? `component docs page`
+        : `automatically generated docs page`;
+      if (betterEntry.name === this.options.docs.defaultName) {
+        logger.warn(
+          `🚨 You have a story for ${betterEntry.title} with the same name as your default docs entry name (${betterEntry.name}), so the docs page is being dropped. Consider changing the story name.`
+        );
       } else {
-        logger.warn(`🚨 Extraction error on ${relativePath}: ${err}`);
-        throw err;
+        logger.warn(
+          `🚨 You have a story for ${betterEntry.title} with the same name as your ${worseDescriptor} (${worseEntry.name}), so the docs page is being dropped. ${changeDocsName}`
+        );
       }
+    } else if (betterEntry.standalone) {
+      // If the other entry was a story entry then this it is better
+      if (worseEntry.type === 'story') return this.chooseDuplicate(worseEntry, betterEntry);
+
+      // Both entries are standalone but pointing at the same place
+      if (worseEntry.standalone) {
+        logger.warn(
+          `🚨 You have two component docs pages with the same name ${betterEntry.title}:${betterEntry.name}. ${changeDocsName}`
+        );
+      }
+      // If one entry is standalone (i.e. .mdx of={}) we are OK with it overriding a template
+      //   - docs page templates, this is totally fine and expected
+      //   - not sure if it is even possible to have a .mdx of={} pointing at a stories.mdx file
+    } else {
+      // If the other entry was a story entry then this it is better
+      if (worseEntry.type === 'story') return this.chooseDuplicate(worseEntry, betterEntry);
```

Comment:
This is kind of a weird recursive structure. I wonder if it makes sense to divide it into two functions: one to choose the better entry & the other to print out the warning/error. It might make the code slightly longer, but I think it would be more clear.

Reply:
OK, that's fair.